### ISSUE TITLE
[IDEA] CheckboxWidget - generalize to also toggle individual items

### ISSUE THREAD

**Issue by twMat** [Jermolene/TiddlyWiki5#5613]
**Is your feature request related to a problem? Please describe.**
The CheckboxWidget is the natural goto for toggling a field value. But, annoyingly, it does not feature toggling of individual items in a field. Unless it is the tags field. So the mechanism *does* exist to toggle individual values... but it just isn't applied for non-tags fields. 

So instead of a direct widget use, one has to construct this and fiddle with *listops* and *​icons*, etc. ​And note that there is no icon that looks like an actual checkbox so if one presents a list with checkboxes to toggle multiple fields (e.g both tags and other fields values) they would get an ugly *mix* of actual checkboxes and surrogate icons.

**Describe the solution you'd like**
I wish the CheckboxWidget could be generalized to toggle any individual value in any field. Possily there could be limitations like "there must only be a single instance of the value in the field".

It would be fine to introduce a **new attribute** to specify that it is a single value that should be toggled and not the whole field value.

**Describe alternatives you've considered**
I did build http://toggle-in-field.tiddlyspot.com/ back in 2018 out of frustration. (BTW, you can see the icon problem there where I've made small png's to resemble checkboxes. Silly!)

**Comment by twMat** [comment]
BTW this has been brought up before but it is never resolved. Clearly there is a need.

e.g #2003 and #2156

**Comment by rmunn** [comment]
There is extensive discussion of this idea in #1957. I'm reading it now, to see if a conclusion was ever reached on how to do it. There seem to be two alternatives that were considered, one of which was extending the CheckboxWidget with a new attribute `tagsFIeld` (though I would say `listField` is better), and the other was a CheckList widget. I don't think they are mutually exclusive — it should be possible to do both — but that's a conclusion I've reached after only skimming through the relevant threads. Maybe there's a comment I missed where someone pointed out a reason to do one over the other.

Also to be considered is the new functionality introduced in 5.1.23 where actions can be tied to checkboxes and radio buttons, and whether that should be leveraged in the implementation. There's a lot to think about before an implementation can be started. I'm interested in implementing this, building on the code written by @matabele in #2003 and #2156, but it'll take a few days before I have finished reading up on the prior work done so that I don't reinvent the wheel. If anyone else already has plans to work on this, please speak up so that we don't duplicate efforts.

**Comment by pmario** [comment]
I'm not sure, but couldn't it be better done using actions and the new "toggle" and "cycle" operators.

I think, if "toggle" in lists is implemented on the widget level the next step is to  implement "cycle". ... and so on.

I think it would be valuable to **try** to improve the documentation of toggle and cycle with the requested functionality. May be it can be implemented that way. 

The advantage is: The user has full control.

just my thoughts

**Comment by saqimtiaz** [comment]
I think it is useful to break down the ability to work with individual titles in a field in two parts:

1. **The ability to toggle items in a field (that is not the tags field)**. While this is currently possible using `<$action-listops>`, it also requires using an additional field to which the checkbox is bound. This is often necessitates using an additional redundant field which otherwise is not needed. It should be possible to invoke actions with `<$checkbox>` without needing to use an additional field.
2. **Specifying the checked/unchecked state.** The checked/unchecked state is currently determined by the value of a field, or in the case of tags the existence of a title in the tags field. For other fields, there is currently no mechanism to tie the state of the checkbox to the presence of a title in a field.

A flexible solution that encompasses more use cases than just toggling a value in a field might be to:
- add support for determining the checked/unchecked state of the checkbox either via the presence of a title in a field, or the result of a filter expression such as `[contains:myfield[myvalue]]` where an empty filter result leads to an unchecked state. Using a filter expression will allow for far more complex logic to be used in determining the state of the checkbox.
- allow the checkbox widget to be used to execute actions without specifying a field that is also toggled on changing the checked state of the checkbox.

I do agree with @pmario that adding very specific support for just toggling a value in a field is not flexible enough.

**Comment by twMat** [comment]
@saqimtiaz wrote:

>support for just toggling a value in a field is not flexible enough.

Well, it would be in *addition* to the existing functionality. What more flexibility would one want from a checkbox widget?

@pmario wrote:

>couldn't it be better done using actions and the new "toggle" and "cycle" operators.

Are (both of) you disqualifying the checkboxwidget all together then? Because also toggling a tag or a fields (full) value could be achieved with individual actions and filter ops. IMO a checkboxwidget serves a basic need to easily toggle (individual) field values.

**Comment by saqimtiaz** [comment]
@twMat suggest re-reading my message carefully. I explain why the current state of things is indeed not good enough for your use case, while also suggesting a way to address your request whilst not restricting the solution to just toggling a value.

**Comment by Jermolene** [comment]
Thanks @twMat @saqimtiaz @rmunn @pmario

I think @saqimtiaz's suggestion is a great approach. Introducing filters judiciously like this can create great opportunities for flexibility.

**Comment by twMat** [comment]
@saqimtiaz - thanks/sorry, you're right, I didn't grasp your points. The idea with a filter to, if I get it right, use the checkbox as a flag sounds cool. And, overall, I'm happy to see there is consensus that the current widget is insufficient.

Maybe irrelevant; but regarding executing actions, it is possible to use the (existing) `actions` attribute for a checkbox that has no `field` argument. There is no default field for the checkbox AFAICT. (No need to reply if this point is irrelevant)

**Comment by rmunn** [comment]
Okay, I think I have an idea of the design. I'll add an attribute that will use a filter to determine the checked/unchecked state of the checkbox. And I think it's *also* a good idea to extend the tag functionality to other fields, because I think it can be done pretty simply. There's just one more question I want to ask for help with before I start working on this, and that is what the attributes should be named.

To save you a trip to tiddlywiki.com to look it up, here are the current attributes of the CheckboxWidget:

Attribute | Description
-- | --
tiddler | Title of the tiddler to manipulate (defaults to the current tiddler)
tag | The name of the tag to which the checkbox is bound
invertTag | When set to yes, flips the tag binding logic so that the absence of the tag causes the checkbox to be checked
field | The name of the field to which the checkbox is bound
index | The index of the tiddler, a DataTiddler, to which the checkbox is bound (be sure to set the tiddler correctly)
checked | The value of the field corresponding to the checkbox being checked
unchecked | The value of the field corresponding to the checkbox being unchecked
default | The default value to use if the field is not defined
class | The class that will be assigned to the label element
actions | A string containing ActionWidgets to be triggered when the status of the checkbox changes (whether it is checked or unchecked)
uncheckactions | A string containing ActionWidgets to be triggered when the checkbox is unchecked
checkactions | A string containing ActionWidgets to be triggered when the checkbox is checked
disabled | Optional, disables the checkbox if set to "yes". Defaults to "no"

I propose adding an attribute called `listField` (`fieldList` would also be a possible name, but I think `listField` works better). It will work like `field`, but will treat that field as a list rather than a single value. The `checked` and `unchecked` attributes will work slightly differently if `listField` is present:

- `checked`, if defined, will be the text searched for in the list.
- `unchecked` will default to empty; if `unchecked` is empty, then the value of the `checked` attribute will be removed when you uncheck the box, and added when you check the box.
- If `unchecked` is *not* empty but `checked` is empty, then it works sort of like `invertTag`: the absence of the `unchecked` value causes the checkbox to be checked, and the value will be removed when you check the box and added if you uncheck the box.
- Finally, if both `unchecked` and `checked` are non-empty, then it works kind of like a toggle: the checkbox will be checked if the `checked` value is present, unchecked if the `unchecked` value is present, and the value of the `default` attribute will be used to determine the checkbox's value if neither is present.
- If `default` is *not* specified and neither value is present, the checkbox will be in an "indeterminate" state using the `checkbox.indeterminate = true` thing that Javascript allows. (It will count as false, so the box will be checked if the user clicks it, but visually it will present as an intermediate state, like a greyed-out box or a box with a minus in it, depending on the browser).

The other attribute I will add is one where I'm not sure of the name, but I'm leaning towards `checkedFilter` or just `filter`. I can see pros and cons to either name, so I'd welcome opinions. That one will be a filter where the filter's result determines what state the checkbox appears to be in. It would interact with the `checked` and `unchecked` attributes as follows:
- neither `checked` nor `unchecked` specified: an empty filter result means unchecked, and a non-empty result (any non-empty result) means checked.
- `checked` specified but `unchecked` omitted or empty: if filter result *contains* the `checked` value (and possibly other values), then the checkbox is checked, otherwise it's unchecked.
- `unchecked` specified but `checked` omitted or empty: works sort of like `invertTag`. If filter result contains the `unchecked` value then the checkbox is unchecked, but if the `unchecked` value is absent then the checkbox is checked.
- `checked` and `unchecked` both specified: works like a toggle. If the filter result contains the `checked` value the checkbox is checked, if it contains the `unchecked` value the checkbox is unchecked, and if neither is present then the `default` attribute is used to determine the checkbox's value.
- If `default` is *not* specified and neither value is present in the filter's result, the checkbox will be in an "indeterminate" state that counts as false, just as with the `listField` attribute.

Note that the rules for how `filter` and `listField` interact with the `checked` and `unchecked` attributes are completely equivalent. So the `listField` attribute could be <del>easily</del> implemented by the `list` filter in combination with actions: `<$checkbox listField="myField">label</$checkbox>` is exactly equivalent to `<$checkbox checkedFilter="[list[!!myField]]" checkactions="something involving action-listops" uncheckactions="something else involving action-listops">label</$checkbox>`. However, I think there's value in providing the `listField` attribute even though the user could implement it independently. Because figuring out the right `listops` string is *not* easy (I'm a professional programmer and yet I punted on figuring it out just now), so I think we want to provide an easy way to do the most common task (make this value appear or disappear in the list) without making people figure out listops actions for the common scenario. If they want more complicated scenarios, the filter and actions attributes will make that possible.

Thoughts? Do you like having the `listField` attribute alongside `filter`? Should I go with a different name for either of those, such as `checkedFilter` or `fieldList`?

**Comment by saqimtiaz** [comment]
@rmunn very happy that you are bringing up the naming of attributes and their interactions. This was the concern in the back of my mind as I wrote my reply yesterday. `<$checkbox>` already has a very large number of attributes and we need to try to make them as easy as possible for users to comprehend and understand. Towards this end, a good approach can sometimes be to try and write the user documentation first. Somewhat similar to what you have already done, but going another step and trying to phrase it in more approachable terms.

I agree that adding an attribute that triggers the field being treated as a list is a good idea. Another approach could be a `mode` attribute that when set to `list`, treats the `field` attribute as a `list` field.

In the interest of simplicity and making the eventual documentation easier to understand, I am not sure that we need the `checkedFilter` to interact with the `checked` and `unchecked` attributes. The use cases for their interaction that you describe can be accomplished by writing the appropriate `checkedFilter` and action strings. I would consider this to be an advanced feature where we can reasonably expect fluency in filters and writing action strings, especially since we would be adding support for list fields separately.

I do prefer `checkedFilter` as the name makes the purpose of the filter clearer.

Also note that it would be useful to pass the new unchecked/checked state of the checkbox as the variable `actionValue` to the actions invoked by the widget. This would make it consistent with the changes in the `<$range>` and `<$radio>` widgets in the last release.

**Comment by pmario** [comment]
> default attribute will be used to determine the checkbox's value if neither is present.

IMO that's not what default is used for in TW. every other widget use default, if the field tiddler doesn't exist. as soon as the tiddler exists, there is a value.

**Comment by rmunn** [comment]
What I mean is that `default` will act more or less the same way it does on the checkbox widget currently: if `checked` is `red` and `unchecked` is `green`, the list `red yellow orange` will produce 🗹 while the list `blue green purple` will produce ☐. But the list `blue yellow octarine`, which contains neither value, will produce 🗹 if `default` is `red` or produce ☐ if `default` is `green`. (And if `default` is neither `red` nor `green` then it will produce an indeterminate checkbox, that will become checked if you click on it).

**Comment by pmario** [comment]
The checkbox widget has many parameters already. I think, we should have a test first, that specifies the status quo.

@Jermolene Is it possible to write tests for the checkbox widget?

**Comment by pmario** [comment]
> What I mean is that default will act more or less the same way it does on the checkbox widget currently: if checked is red and unchecked is green, the list red yellow orange will produce 🗹 while the list blue green purple will produce ☐. But the list blue yellow octarine, which contains neither value, will produce 🗹 if default is red or produce ☐ if default is green. (And if default is neither red nor green then it will produce an indeterminate checkbox, that will become checked if you click on it).

That's not how default is used atm. Look at the code! default is only active, if the tiddler or the field that it is bound to doesn't exist. This behaviour is consistently used in TW for the default value.

**Comment by pmario** [comment]
>What I mean is that default will act more or less the same way it does on the checkbox widget currently ..

I think it has to work exactly the same way, otherwise we have a compatibility problem

**Comment by pmario** [comment]
@rmunn I'm using ubuntu atm. There are some unreadable characters in your post. may be the windows pc will show them. later today.

**Comment by rmunn** [comment]
@saqimtiaz -

> Somewhat similar to what you have already done, but going another step and trying to phrase it in more approachable terms.

Yeah, I was definitely going for technical documentation rather than user-centric when I wrote that list of bullet points. It's basically an algorithm description. I'd use non-technical language with examples in user-centric documentation.

> I am not sure that we need the `checkedFilter` to interact with the `checked` and `unchecked` attributes.

I did that because I thought I needed it to work in parallel with how the `listField` attribute worked, but now that I think about it, if you have `checkedFilter` without `listField` then there's no way to know which field to add/remove the text from. So you're right, they shouldn't. If you're using `checkedFilter` then you need to write something with listops. If you want the simple case (add/remove a value, or toggle a value) then you use `listField` and `checked`/`unchecked`.

> I do prefer checkedFilter as the name makes the purpose of the filter clearer.

I think the drawback is confusion between the `checked` and `checkedFilter` names, which do something rather different. If `checkedFilter` is interacting with `checked` then it's easier to understand, but you're right that it shouldn't interact. So that leaves me a bit worried that users will say, "Huh? What's the difference between `checked` and `checkedFilter` and when should I use one over the other?" Which led me to prefer the name `filter` as it doesn't have the word "checked" in it.

> Also note that it would be useful to pass the new unchecked/checked state of the checkbox as the variable `actionValue` to the actions invoked by the widget. This would make it consistent with the changes in the `<$range>` and `<$radio>` widgets in the last release.

Good point. I'll make sure to look at the range and radio widget code as I implement this, and pass in the state similarly. The behavior *should* be the same between those widgets.

**Comment by rmunn** [comment]
@pmario -

> @rmunn I'm using ubuntu atm. There are some unreadable characters in your post. may be the windows pc will show them. later today.

I'll double-check that I was using UTF-8, but I'm on Ubuntu as well, where everything defaults to UTF-8 (as everything should) so the unreadable characters should be readable to you. Perhaps you don't have a font installed with emoji in it? The 🗹 character is U+1F5F9, and should appear as a box with a checkmark in it. The ☐ character is U+2610, and should appear as a plain box without a checkmark in it.

I'll address the rest of your comments, which came in while I was writing my reply to saqimtiaz, soon. Just wanted to get this one written first so you'll know what characters I was writing there.

**Comment by rmunn** [comment]
The way `default` works in the checkbox widget is it should be a value that matches either the value of `checked` or the value of `unchecked`, and is used if the field is undefined so there's no value to be found. E.g.,

```
<$checkbox field="status" checked="open" unchecked="closed" default="closed"> Is it open?</$checkbox>
```

So what's the *purpose* behind that? The purpose is that the first time the wiki is created, the field will probably not exist, and you want to define what's the first state of the checkbox. With the filter, it's far less likely to happen that the field doesn't exist. But it's *much* more likely to happen that a list field contains neither the `checked` nor `unchecked` value. So what value should the checkbox show if that happens? How do you decide that value?

Putting myself in the shoes of a user, I'd expect `default` to be the way to decide that value. I'd expect `default` on a checkbox to mean "if you can't tell whether it's true or false, then make it true". If I had the list `blue yellow octarine`, and defined `default` to be either `red` or `green`, I'd *want* `default` to be used to determine the value since neither `red` nor `green` was present in the list. I would *not* want it to just assume that I wanted the checkbox to be unchecked if neither value was present, because there might be times when I want checked to be the default. And therefore, if I was a user and `default` did *not* work that way, I'd be filing a bug report.

I understand your point about this being a change in the meaning of `default` from how it's used elsewhere. But if I don't use `default`, I have to give the user *some* way of specifying the behavior when neither value is present in the filter result. What would you suggest I do? The only alternative I can see is adding a third attribute, and (as you point out in https://github.com/Jermolene/TiddlyWiki5/issues/5613#issuecomment-823041795) the checkbox widget is already groaning under a huge pile of attributes. I'm already adding two of them; I don't want to add a third when there's already a perfectly good one that users will expect to use for this purpose.

**Comment by saqimtiaz** [comment]
> I think the drawback is confusion between the `checked` and `checkedFilter` names, which do something rather different.

I actually prefer that the names are similar so they can be listed together and then we document that you use one or the other, and which one takes preference. Both are ways to determine whether the checkbox should be checked or not.

From docs:
> Checked The value of the field corresponding to the checkbox being checked

To which we could add something like:
`checkedFilter: takes precedence over Checked if both are specified. A non empty filter result corresponds to the checkbox being checked.`

 I am open to other names but `filter` seems far too unspecific.

**Comment by Jermolene** [comment]
> The checkbox widget has many parameters already. I think, we should have a test first, that specifies the status quo.
> 
> 
> 
> @Jermolene Is it possible to write tests for the checkbox widget?

Hi @pmario it is currently possible to write tests for the initial rendering and subsequent refreshing of widgets, but not for interactive functionality.

**Comment by rmunn** [comment]
> ... it is currently possible to write tests for the initial rendering and subsequent refreshing of widgets, but not for interactive functionality.

I will probably try to set something up with https://www.cypress.io/ to test the interactive functionality. Set up a TW page to be loaded in a Cypress test environment and have the fields visible. Check and uncheck the box and make sure the field's contents remain what they should be. If that works, I'll try to describe my test setup, because the same technique could be applied to testing quite a lot of other interactive features in TW.

**Comment by pmario** [comment]
ubuntu had a problem with: 🗹 ... Window can see it

**Comment by pmario** [comment]
>Hi @pmario it is currently possible to write tests for the initial rendering and subsequent refreshing of widgets, but not for interactive functionality.

Yea, I did have a look at the test tiddlers, that test widgets. ... So we can check for the rendering status, if the different parameters exist and if the "stat-tiddler / state-fields" exist. ... So we see the "init-status" .... 

IMO this has to do for the beginning.

**Comment by pmario** [comment]
>I did that because I thought I needed it to work in parallel with how the listField attribute worked, but now that I think about it, if you have checkedFilter without listField then there's no way to know which field to add/remove the text from. So you're right, they shouldn't. If you're using checkedFilter then you need to write something with listops. If you want the simple case (add/remove a value, or toggle a value) then you use listField and checked/unchecked.

@rmunn I think, it's important to exactly describe, which parameter takes precedence. So users can be sure about the functionality. 

As Saq wrote. **We should definitely write the "user documentation" for the new functionality first.** ... If we have problems to explain it properly, the algorithm may have a problem too.

**Comment by pmario** [comment]
>..So that leaves me a bit worried that users will say, "Huh? What's the difference between checked and checkedFilter and when should I use one over the other?" Which led me to prefer the name filter as it doesn't have the word "checked" in it.

I think, if the "which parameter takes precedence" is clear, this may be much easier. Also in the logic later

**Comment by pmario** [comment]
> Putting myself in the shoes of a user, I'd expect `default` to be the way to decide that value. I'd expect `default` on a checkbox to mean "if you can't tell whether it's true or false, then make it true".

Yea, but making it true, doesn't mean the checkbox is checked. It means the logic that is executed is true. 

The important thing is. We need to have tests first. The existing behaviour can't be changed for the existing params.

For the new params we need to make it work in a useful and sensible way. 

> If I had the list `blue yellow octarine`, and defined `default` to be either `red` or `green`, I'd _want_ `default` to be used to determine the value since neither `red` nor `green` was present in the list. I would _not_ want it to just assume that I wanted the checkbox to be unchecked if neither value was present, because there might be times when I want checked to be the default. And therefore, if I was a user and `default` did _not_ work that way, I'd be filing a bug report.

The check box needs to be defined, if we can be 100% sure about the state. If `default` is `red` and the list is `green` ... **The list field exists** and therefore, the **default is useless**. The decision has to be made by the `filter`

**Comment by pmario** [comment]
> The check box needs to be defined, if we can be 100% sure about the state. If `default` is `red` and the list is `green` ... **The list field exists** and therefore, the **default is useless**. The decision has to be made by the `filter`

The filter is defined by the user. So the user in control, that's important.

**Comment by pmario** [comment]
I think the least confusing thing is, if the list-handling works similar to the tag-handling. 

List toggling is the same as "add-tag" / "remove-tag"

**Comment by rmunn** [comment]
> As Saq wrote. **We should definitely write the "user documentation" for the new functionality first.** ... If we have problems to explain it properly, the algorithm may have a problem too.

Yes. I plan to write the documentation first, then automated tests for the old functionality and the new functionality (the tests for the new functionality will fail, of course), and then write the actual code. So TDD (test-driven development) with an added helping of DDD (documentation-driven development).

And I'll make sure to explain which parameter has precedence in cases where there might be a conflict (e.g., if both `field` and `listField` exist, which one should win?). Actually, if `field` and `listField` are both present, I haven't figured out yet which one should win, so that's a case where I could still use some suggestions.

> > The check box needs to be defined, if we can be 100% sure about the state. If `default` is `red` and the list is `green` ... **The list field exists** and therefore, the **default is useless**. The decision has to be made by the `filter`
> 
> The filter is defined by the user. So the user in control, that's important.

So what you're saying is, we should tell people "make sure you write a filter that turns into a known state". I agree that that's a good idea, but I do want to allow the very simple use case to stay simple. And the *simple* use case is the one where you're not writing a filter, but just using `listField`. The user sets `listField` equal to "colors", and the colors list contains `blue purple`. This particular checkbox has `red` and `green` for its checked and unchecked states. How do you pick the checkbox state? Remember, this is the *simple* use case where the user has *not* used the `filter` attribute. So far, for this scenario, I haven't come up with any better solution than to use the `default` attribute, and expand it just a little. Instead of having the meaning of "default if the field doesn't exist", I have it being "default if neither yes nor no is *found*". But only for the `listField` functionality; I wouldn't change how it works for the `field` functionality. I think making simple things simple for the usres is a more important design principle than making everything 100% consistent. If consistency in one spot means adding user complexity, then a little inconsistency in that spot is better.

Ultimately, it's @Jermolene's project, and he can make the call. I'll write it the way I think it should work, but if he says it needs to be the other way, I'll change it.

**Comment by saqimtiaz** [comment]
> Yes. I plan to write the documentation first, then automated tests for the old functionality and the new functionality (the tests for the new functionality will fail, of course), and then write the actual code. So TDD (test-driven development) with an added helping of DDD (documentation-driven development).

Sounds great, thank you!

> Actually, if `field` and `listField` are both present, I haven't figured out yet which one should win, so that's a case where I could still use some suggestions.

This is why I am wondering if a `mode` or `fieldMode` attribute that can be set to list for list handling might be better. Then users know to always use the `field` attribute to specify which field to use, and the `fieldMode` attribute to trigger it to optionally be treated as a list field. Perhaps see if this makes any sense when you try to write out the docs @rmunn

**Comment by joshuafontany** [comment]
Following with interest because I plan on using the Checkbox Widget quite a
lot for UI...

How complicated would it be to just interpret the `checked` and `unchecked`
attributes AS filters.

Then a simple list="yes" attribute to toggle the list behaviour. This would
allow toggling the result of the checked/unchecked results as a list or as
a single string item.

`tag="foo"` would be a deprecated/quick way of saying `list="yes"
field="tags" checked="foo"` and
`tag={{{ [<foo>match<bar>else[fnord]] }}}` could be expanded to
`list="yes" field="tags" checked="[<foo>match<bar>else[fnord]]"`
AND also allow targeting the same logic at list Indexes (currently not
possible).

Thoughts? Best,

Joshua F

**Comment by rmunn** [comment]
@joshuafontany wrote:
> How complicated would it be to just interpret the `checked` and `unchecked` attributes AS filters. Then a simple list="yes" attribute to toggle the list behaviour. This would allow toggling the result of the checked/unchecked results as a list or as a single string item.

I've thought about this idea, and there's one part of it where I don't know how it would work. As you know, attributes can be set with `attr={{{ filter }}}` to get a single string value (the first item of the filter if it returns multiple items), so if all you want is for `checked` and `unchecked` to be single items set by a filter then the code can already do that. So I'm guessing that you want to be able to set multiple items in `checked` and/or `unchecked`. I can't quite figure out how that would work. If the `checked` filter evaluates to `red green` (that is, two items, one "red" and one "green"), then what is the rule for whether the checkbox is checked or not? Is it "the list contains *either* 'red' *or* 'green' "? Or is it "the list contains *both* 'red' *and* 'green' "? And then there's the question of what happens if the list contains *some* of the items in the `checked` list and also *some* of the items in the `unchecked` list. What should the checkbox state be then? It gets complicated quickly.

I think in the end, I'm going to decide against having `checked` and `unchecked` be allowed to be multiple items, and require that they be single strings (or absent/empty). Which means that if you want them to be assignable via a filter, you'd just write `checked={{{ [<foo>match<bar>else[fnord]] }}}`, and you'd just have to make sure that the `<foo>match<bar>` filter returns just one item, or that the item you care about is the first item in the list that it returns.

You also mentioned "targeting the same logic at list indexes" not being possible currently, but I only have a vague idea what you mean. I'll explain what I *think* you're asking for; could you please confirm whether I've gotten it right, or if not, expand on that idea a little? My guess is that you'd like something to say "This checkbox should set/unset the item at position 1 in this list", i.e. if the list is `red green blue` you want to be able to say `<$checkbox field="colors" fieldMode="list" listIndex=1/>` and that would produce a checkbox that targets the word "green". But what happens when the user turns that checkbox off? Now the list contains `red blue` and listIndex 1 means "blue". The meaning of the checkbox changes. Do you know what it should do in that case?

Now, I *can* think of a way to target list indexes (if I've understood you correctly) without adding a `listIndex` attribute, by using the `{{{ filter }}}` syntax. You'd want one list that's the "target" list containing only the values that are turned on by checkboxes, and another list that's the "source" list containing all the possible values. Let's call the source list "colors" and say it has "red green blue" in it, and the target list "currentColors". Then to target index 1 (green), you'd write:

```js
<$checkbox field="currentColors" fieldMode="list" checked={{{ [list[!!colors]nth[1]] }}} />
<!-- TODO: Put the name in there as a checkbox label -->
```

And that could be part of a `<$list>` widget that either uses `<<index>>` to get the index that goes into the `checked` filter, or else you could just do this that's simpler:

```js
<$list variable="color" filter="[list[!!colors]]">
<$checkbox field="currentColors" fieldMode="list" checked=<<color>> ><<color>></$checkbox>
</$list>
```

Is that what you meant by targeting list indexes? Or did you mean something else, that my example doesn't cover?

**Comment by rmunn** [comment]
I've written the documentation for field lists in #5623. Haven't yet written the documentation for filter mode; that will come later. I used @saqimtiaz 's suggestion of `fieldMode="list"`. If you have comments or suggestions to improve the documentation, please leave feedback on #5623.

**Comment by rmunn** [comment]
In a review on the PR, @saqimtiaz has suggested `fieldType="list"` as a possible alternative to `fieldMode="list"`. I'd like to get feedback from the people most interested in this feature. Between `<$checkbox field="foo" fieldMode="list" />` or `<$checkbox field="foo" fieldType="list" />`, which one looks/feels better to you?

**Comment by joshuafontany** [comment]
Really excellent response, I agree. To clarify, by "targeting indexes" I meant indexes-of-data-tiddkers as opposed to fields-of-regular-tiddlers. As such, maybe just "mode" instead of "fieldMode"?

**Comment by rmunn** [comment]
@joshuafontany - So you're saying that you want to be able to have a data tiddler with indexes "a", "b", "c" and so on, and you'd like the checkbox to be able to treat index "a" of the data tiddler as a list, just like it can (with `fieldMode="list"`) treat the field of the tiddler as a list. Am I understanding your request now?

**Comment by saqimtiaz** [comment]
@rmunn @joshuafontany could this use case not be accommodated by the use of action strings and the proposed checkFilter attribute?

The idea behind that proposal was to have the flexibility to not have to explicitly code for other use cases and avoid adding more complexity to the widget.

**Comment by rmunn** [comment]
@saqimtiaz If the attribute is called `mode` instead of `fieldMode` then it doesn't add any extra complexity (at least, from a user-documentation point of view) to treat indexes of data tiddlers equally with fields of regular tiddlers. I never use data tiddlers myself, but the intent in TW seems to be that they are supposed to be on an equal footing with fields, with everything that works on fields *supposed* to work on data tiddlers as well.

Yes, the list manipulation in a data tiddler could be done via filters and actions, but that would make fields "first-class citizens" and data tiddlers "second-class citizens" for this new list feature. And if I can let both fields and indexes participate equally in the feature without adding conceptual complexity, I'll do it. By making the attribute just called `mode` instead of `fieldMode` it allows both uses without adding any complexity.

I'll update my documentation to refer to a "mode" attribute and then you can give feedback on whether I've made the documentation too complex.

**Comment by saqimtiaz** [comment]
@rmunn I've been under the impression that we weren't intending to actively promote or encourage usage of data tiddlers by adding more affordances. 

I may be wrong on that front, perhaps @Jermolene can clarify.

**Comment by twMat** [comment]
>I've been under the impression that we weren't intending to actively promote or encourage usage of data tiddlers by adding more affordances.

[Off topic] If a *better* alternative than data tiddlers comes up, *then* it makes sense to disencourage use of data tiddlers. So far, data tiddlers (and their superior simplicity in use) is the best we have. Consequently, implementations that need the *functionality* provided by data tiddlers, should use data tiddlers. But of course, it would make sense to keep those particular parts in the implementation as isolated as possible so they can be easily exchanged when the better alternative comes, one sunny day.

**Comment by pmario** [comment]
>By making the attribute just called `mode` instead of `fieldMode` it allows both uses without adding any complexity.

I think that's sensible, it makes the "term" more generic and it doesn't add complexity.

**Comment by saqimtiaz** [comment]
@rmunn #4749 might be worth a look as you work on this.

**Comment by joshuafontany** [comment]
@rmunn Yes, keeping the functions in parity no matter the target type. I tend to use JSON Data tiddlers heavily for certain things, and I like where these changes are headed. Thanks!

**Comment by rmunn** [comment]
As I work on this, I'm finding that I want to change my design a little. The `fieldMode` attribute is now called `mode`, because I'm making list mode work with both indexes of data tiddlers and with fields of normal tiddlers. So if `mode="list"` then either the `field` or `index` attribute is used to determine where the checkbox's state should be stored, and the `checked` and `unchecked` attributes are used to determine what gets stored in the state. For filter mode, instead of using a `checkedFilter` attribute, you just use `mode="filter"`. And then the `checked` and `unchecked` attributes are treated as filters instead of as simple strings.

The way `checked` and `unchecked` work in list mode will still be the same as what I wrote in https://github.com/Jermolene/TiddlyWiki5/issues/5613#issuecomment-823015596, so I won't duplicate that here. But in filter mode, the way `checked` and `unchecked` work is as follows:

* If `checked` is present and `unchecked` is absent, then the `checked` attribute is evaluated as a filter. If its result is non-empty, the checkbox is checked; if the filter result is empty, the checkbox is unchecked.
* If `unchecked` is present and `checked` is absent, then the `unchecked` attribute is evaluated as a filter. If its result is empty, the checkbox is checked; if the filter result is non-empty, the checkbox is unchecked.
* If both `checked` and `unchecked` are present, then they are both evaluated. If just one of them evaluates to a non-empty result, then the checkbox is either checked or unchecked. If both of them evaluate to non-empty, or if both evaluate to empty, then the checkbox state is indeterminate. (An indeterminate state is treated internally as unchecked, so that if an indeterminate checkbox is clicked on, the `checkactions` will fire).
    * Possible advanced feature, up for debate: should indeterminate states have two possible internal values, like "indeterminate - true" and "indeterminate - false"? The rationale would be that filters are for advanced users, who might have reason to want to treat indeterminate checkboxes as being "false" (will fire `checkactions` when clicked on) or "true" (will fire `uncheckactions` when clicked on) depending on their use case. If we go for this advanced feature, then when both the `checked` and `unchecked` filters return non-empty, the checkbox is in an "indeterminate - true" state, but when both filters return empty values, then the checkbox is in an "indeterminate - false" state. I'm not sure this has a use case, but it might, and it would be easy to implement.
* The `default` attribute is ignored in filter mode, as there is no field or index whose value is being checked. Any defaults desired by the user can be written into the `checked` and `unchecked` filters.

I'd welcome comments on this redesign, where the `mode` attribute can be either "list", "filter", or empty. That way I avoid having to add a `checkedFilter` attribute, and the entire question of how the `checked` and `unchecked` attributes interact with filter mode (answer: they *become* the `checkedFilter` attribute, and we allow for an `uncheckedFilter` attribute at no extra cost).

**Comment by pmario** [comment]
>Possible advanced feature, up for debate: should indeterminate states have two possible internal values, like "indeterminate - true" and "indeterminate - false"? 

I would go with the pragraph above. It "sounds" OK. ... I think the undetermined state isn't used. ... If advanced users want to use filters, they should make sure that the filters work as they want them to work. IMO no extra state in the widget should be needed. 

I think where ever you had problems in the documentation to describe the behaviour in a sensible way, there may be a problem in the logic. ... 

If everything is "easy" .. Just program it that way. 

just a thought.

**Comment by saqimtiaz** [comment]
@rmunn that sounds reasonable. If you run into any further uncertainty while working on this, I find that writing out examples of how the widget would be used for different scenarios helps (and later forms part of the examples in the documentation).

**Comment by twMat** [comment]
What is needed to resolve this issue? Anyone?

**Comment by saqimtiaz** [comment]
@twMat presumably @rmunn is otherwise occupied. So it needs a developer with the availability and interest in implementing this.

**Comment by twMat** [comment]
>@twMat presumably @rmunn is otherwise occupied. So it needs a developer with the availability and interest in implementing this.

Yes, thanks, that's understood. My question was more if some difficulty in implementing it has been identified and could be formulated here, so to simplify for anyone wanting to pick it up.

**Comment by rmunn** [comment]
Sorry for my long silence. I've been busy with other projects and haven't found time to work on TW recently. But I should have some time in the next couple of weeks, and I'm excited to get back to work on this issue.

My thinking on the design, though, is back to my original idea of adding a `listField` attribute rather than a `mode` attribute. And the "filter" mode would be implemented by a `filter` attribute (rather than `checkedFilter`). Then the `checked` and `unchecked` attributes would remain unchanged: they define the value that should be in the field, or the list, or the filter result. I'm running out of time today to be able to write more details about my current thinking, but I should have more time for this soon.

**Comment by twMat** [comment]
@rmunn - thanks for picking it up again! It will make for a very useful improvement :-)

**Comment by rmunn** [comment]
#6561 is ready to review. Once it's merged, I'll start working on #6564.

**Comment by rmunn** [comment]
Now that this has been implemented, we can probably close #1957 as well.
